lib/internal/util.js

```diff
@@ -36,6 +37,35 @@ function objectToString(o) {
 // each one once.
 const codesWarned = {};
 
+// Returns a function that emits a deprecation on first run
+// For internal use only. `code` is a required argument.
+// If --no-deprecation is set, then it is a no-op.
+// { pending: true } are visible only when pendingDeprecation is on
+function deprecator(msg, code, options = {}) {
+  if (process.noDeprecation === true) {
+    return () => {};
+  }
+  if (options.pending && !pendingDeprecation) {
+    return () => {};
+  }
+
+  // code should be set
+  if (typeof code !== 'string' || code === '')
+    throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'code', 'string');
+
+  let warned = false;
+
+  function deprecation(caller) {
+    if (warned) return;
+    warned = true;
+    if (codesWarned[code]) return;
+    codesWarned[code] = true;
```

Comment:
As question: is `warned` here to prevent a lookup in the default case?

Reply:
Yes.
1. If this specific function has already been called (which it would be in most cases) — `warned` will be `true`, and this will return asap.
2. The second check (`codesWarned[code]`) is needed to catch the situation where e.g. _another_ function has been created using the same deprecation code (or the warning was emitted through `deprecate` below) — we still don't want to display it second time in that case.